ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +175,424 @@ let order_f (_, vm_rec) =
   let main, shadow =
     Memory_check.vm_compute_start_memory ~__context ~policy snapshot
   in
   Int64.add main shadow
 
-(** Return a VM -> Host plan for the Host.evacuate code. We assume the VMs are all agile. The returned plan may
-    be incomplete if there was not enough memory. *)
-let compute_evacuation_plan ~__context total_hosts remaining_hosts
-    vms_and_snapshots =
-  let hosts =
-    List.map
-      (fun host ->
-        ( host
-        , Memory_check.host_compute_free_memory_with_maximum_compression
-            ~__context ~host None
-        )
-      )
-      remaining_hosts
+let host_free_memory ~__context ~host =
+  Memory_check.host_compute_free_memory_with_maximum_compression ~__context
+    ~host None
+
+let vm_memory ~__context snapshot =
+  let policy =
+    match Helpers.check_domain_type snapshot.API.vM_domain_type with
+    | `hvm | `pv ->
+        Memory_check.Dynamic_min
+    | `pv_in_pvh | `pvh ->
+        Memory_check.Static_max
   in
-  let vms =
-    List.map
-      (fun (vm, snapshot) ->
-        let policy =
-          match Helpers.check_domain_type snapshot.API.vM_domain_type with
-          | `hvm | `pv ->
-              Memory_check.Dynamic_min
-          | `pv_in_pvh | `pvh ->
-              Memory_check.Static_max
+  total_memory_of_vm ~__context policy snapshot
+
+module VMGrpMap = Map.Make (struct
+  type t = [`VM_group] Ref.t
+
+  let compare = Ref.compare
+end)
+
+module HostKey = struct
+  type t = [`host] Ref.t
+
+  let compare = Ref.compare
+end
+
+(* For a VM anti-affinity group, the state of a host which determines
+   evacuation planning for anti-affinity VMs in that group:
+   1. vm_cnt: the number of running VMs in that group resident on the host
+   2. h_size: the amount of free memory of the host *)
+module AntiAffinityEvacPlanHostState = struct
+  type t = int * int64
+
+  (* During evacuation planning for anti-affintiy VMs, "vm_cnt" is the first
+     factor considered, "h_size" is the second factor considered:
+     Let's say the next anti-affinity VM to be planned belongs to group A, the
+     host to be selected should be the one which has minimal "vm_cnt" of group
+     A, for hosts with the same "vm_cnt", pick the one with the minimal
+     "h_size" which can hold the VM(h_size >= vm_size). *)
+  let compare (vm_cnt_0, h_size_0) (vm_cnt_1, h_size_1) =
+    match Int.compare vm_cnt_0 vm_cnt_1 with
+    | 0 ->
+        Int64.compare h_size_0 h_size_1
+    | c ->
+        c
+end
+
+(* A Psq of hosts for an anti-affinity group, which is used for evacuation
+   planning for anti-affinity VMs in that group: the minimal host in this Psq
+   is the first one to be considered to plan a VM from that group.
+   When several hosts share the minimal "AntiAffinityEvacPlanHostState",
+   the minimal host is the host with the smallest ref. *)
+module AntiAffinityEvacPlanHostPsq =
+  Psq.Make (HostKey) (AntiAffinityEvacPlanHostState)
+module Q = AntiAffinityEvacPlanHostPsq
```

Comment:
Move it into where it is used

Reply:
Or shorten the names of `AntiAffinityEvacPlanHostPsq` and `AntiAffinityEvacPlanHostState` directly.
How about the following?
`AntiAffinityEvacPlanHostState` -> `HostPlanState`
`AntiAffinityEvacPlanHostPsq` -> `PoolPlanState`